libraries/AP_GPS/AP_GPS_SBF.cpp

```diff
@@ -0,0 +1,368 @@
+            sbf_msg.sbf_state = sbf_msg_parser_t::LENGTH2;
+            break;
+        case sbf_msg_parser_t::LENGTH2:
+            sbf_msg.length += (uint16_t)(temp << 8);
+            sbf_msg.sbf_state = sbf_msg_parser_t::DATA;
+            if (sbf_msg.length % 4 != 0) {
+                sbf_msg.sbf_state = sbf_msg_parser_t::PREAMBLE1;
+                Debug("bad packet length=%u\n", (unsigned)sbf_msg.length);
+            }
+            break;
+        case sbf_msg_parser_t::DATA:
+            if (sbf_msg.read >= sizeof(sbf_msg.data)) {
+                Debug("parse overflow length=%u\n", (unsigned)sbf_msg.read);
+                sbf_msg.sbf_state = sbf_msg_parser_t::PREAMBLE1;
+                break;
+            }
+            sbf_msg.data[sbf_msg.read] = temp;
+            sbf_msg.read++;
+            if (sbf_msg.read >= (sbf_msg.length - 8)) {
+                uint16_t crc = crc16_ccitt((uint8_t*)&sbf_msg.blockid, 2, 0);
+                crc = crc16_ccitt((uint8_t*)&sbf_msg.length, 2, crc);
+                crc = crc16_ccitt((uint8_t*)&sbf_msg.data, sbf_msg.length - 8, crc);
+
+                sbf_msg.sbf_state = sbf_msg_parser_t::PREAMBLE1;
+
+                if (sbf_msg.crc == crc) {
+                    return process_message();
+                } else {
+                    Debug("crc fail\n");
+                    crc_error_counter++;
+                }
+            }
+            break;
+    }
+
+    return false;
+}
+
+void
+AP_GPS_SBF::log_ExtEventPVTGeodetic(msg4007* temp)
+{
+    if (gps._DataFlash == NULL || !gps._DataFlash->logging_started()) {
+        return;
+    }
+
+    uint64_t now = hal.scheduler->micros64();
+
+    struct log_GPS_SBF_EVENT header = {
+        LOG_PACKET_HEADER_INIT(LOG_GPS_SBF_EVENT_MSG),
+        time_us:now,
+        TOW:temp->TOW,
+        WNc:temp->WNc,
+        Mode:temp->Mode,
+        Error:temp->Error,
+        Latitude:ToDeg(temp->Latitude),
+        Longitude:ToDeg(temp->Longitude),
+        Height:temp->Height,
+        Undulation:temp->Undulation,
+        Vn:temp->Vn,
+        Ve:temp->Ve,
+        Vu:temp->Vu,
+        COG:temp->COG
+    };
+
+    gps._DataFlash->WriteBlock(&header, sizeof(header));
+}
+
+bool
+AP_GPS_SBF::process_message(void)
+{
+    uint16_t blockid = (sbf_msg.blockid & 4095u);
+
+    Debug("BlockID %d", blockid);
+
+    // MeasEpoch
+    if (blockid == 4027) {
+
+    }
+    // ExtEventPVTGeodetic
+    if (blockid == 4038) {
+        log_ExtEventPVTGeodetic((msg4007*)sbf_msg.data);
+    }
+    // PVTGeodetic
+    if (blockid == 4007) {
+        msg4007* temp = (msg4007*)sbf_msg.data;
+
+        // Update time state
+        if (temp->WNc != 65535) {
+            state.time_week = temp->WNc;
+            state.time_week_ms = (uint32_t)(temp->TOW);
+        }
+		
+		state.last_gps_time_ms = hal.scheduler->millis();
+
+        state.hdop = last_hdop;
+
+        // Update velocity state (dont use −2·10^10)
+        if (temp->Vn > -20000000000) {

```

Comment:
i think we should mark large integer constants with LL to be a "long long" constant, otherwise the result depends on the compiler